Relevant files for useWorkshopContext

faye-tan opened the pull request:
### How It Works Diagram
![bidirectional-iframe-diagram (3)](https://github.com/user-attachments/assets/7dbd06a3-c526-42d7-aed0-7c7bf75c0401)

### What's included
- The entrypoint is `useWorkshopContext.ts` which takes in a config definition and returns `IAsyncLoaded<WorkshopContext>`
- `WorkshopContext` is a mapped type that takes `IConfigDefinition` which is a list of config field definitions, and transforms it into a config object strongly typed with fieldIds as property names, and field variableType as the property value type
- See `Examples.tsx` for example usage of the hook, and `ExampleConfig.ts` for an comprehensive example of all config field types 


### Follow ups
- Holding off on struct support, as we do not have the ability in Workshop to have variable pickers for structs that will enforce "strict equality". For example, if struct `personStruct` in the config requires fields `{ name: string; birthdate: Date, allergies: string[] }`, the Workshop struct variable picker should filter for struct variables that are a superset of those fields. Otherwise passing it to the plugin could have averse effects. 

### Next steps
- Follow steps in https://github.palantir.build/devtools/open-source to get this repo published as an npm package under the public [Palantir github org](https://github.com/palantir)

mduczek commented on src/useWorkshopContext.ts:
```diff
@@ -0,0 +1,107 @@
+    /**
+     * Only receives a message of type IWorkshopReceivedConfigMessage. 
+     */
+    const handleWorkshopReceivedConfigMessage = (_event: MessageEvent<IWorkshopReceivedConfigMessage>) => {
+        createConfigFromConfigFields(configFields.fields); 
```
how do we make sure we only accept messages from Workshop and no one else

mduczek commented on src/useWorkshopContext.ts:
```diff
@@ -0,0 +1,107 @@
+     */
+    React.useEffect(() => {
+        sendConfigDefinitionToWorkshop(configFields);  // TODO what if configFields input changes?? 
+        initializeListeners();
```
would be good to document the flow of messages:
1. Workshop mounts the iframe and Workshop sends a message to the iframe using the ref
2. the iframe sends the message back to the original sender (Workshop)
3. the iframe only accepts the messages from the original sender (Workshop)

mduczek commented on src/useWorkshopContext.ts:
```diff
@@ -0,0 +1,107 @@
+    /**
+     * Convert the given config definition into config layers, filled in with default values. 
+     */
+    const createConfigFromConfigFields = (configFields: IConfigDefinitionFields["fields"]) => {
```
```suggestion
    const createConfigFromDefinition = (configFields: IConfigDefinitionFields["fields"]) => {
```

mduczek commented on src/useWorkshopContext.ts:
```diff
@@ -0,0 +1,107 @@
+                            outputValueTypes[fieldId] = field.fieldType.outputVariableType;
+                            return; 
+                    }
+                case "listOf": 
+                    listOfConfigs[fieldId] = Array(field.maxLength).fill(createConfigFromConfigFields(field.config));
+                    return; 
```
nit add assertNever

mduczek commented on src/useWorkshopContext.ts:
```diff
@@ -0,0 +1,107 @@
+        const outputValueTypes: { [id: string]: IVariableType } = {}; 
+        const listOfConfigs: { [id: string]: ContextConfigLayer[] } = {};
+
+        Object.entries(configFields).forEach(([fieldId, field]) => {
```
```suggestion
        forEach(configFields, (field, fieldId) => {
```

mduczek commented on src/utils.ts:
```diff
@@ -0,0 +1,75 @@
+ * Sends a message to Workshop through the window. 
+ */
+export function sendMessageToWorkshop(message: IMessageToWorkshop) {
+    window.postMessage(JSON.stringify(message), "*");
```
we should say why it's fine to send the message everywhere

mduczek commented on src/utils.ts:
```diff
@@ -0,0 +1,75 @@
+export function isStruct(structType: IVariableType_Struct, value?: VariableValue): value is StructValue {
+    if (value != null && value["structFields"] != null && typeof value["structFields"] === "object") {
+        return Object.entries(value["structFields"]).every(([structFieldId, structFieldValue]) => {
+            if (!isString(structFieldId)) {
```
keys in an object are always string
<img width="258" alt="Screenshot 2024-08-30 at 11 16 03 PM" src="https://github.com/user-attachments/assets/44b86616-465c-41d0-83de-ebc3bc1e6406">

mduczek commented on src/utils.ts:
```diff
@@ -0,0 +1,75 @@
+            return false; 
+        case "struct": 
+            return isStruct(variableType, value); 
+        case "objectSet":
```
add assertNever

mduczek commented on src/types/messages.ts:
```diff
@@ -0,0 +1,90 @@
+ * Messages to send to Workshop.
+ */
+export type IMessageToWorkshop = 
+    ISendingConfigToWorkshopMessage | 
+    ISettingWorkshopValue | 
+    IExecutingWorkshopEvent;
```
```suggestion
    ISendConfigToWorkshopMessage | 
    ISetWorkshopValue | 
    IExecuteWorkshopEvent;
```

mduczek commented on src/types/messages.ts:
```diff
@@ -0,0 +1,90 @@
+    type: MESSAGE_TYPES_FROM_WORKSHOP.CONFIG_RECIEVED; 
+}
+        
+export interface IValueChangeFromWorkshopMessage {
```
should we pass the whole config for now in a message?
I think this will simplify things

mduczek commented on src/example/Example.tsx:
```diff
@@ -0,0 +1,47 @@
+
+    const setValue = React.useCallback(() => {
+        if (isAsyncStatusLoaded(workshopContext)) {
+            workshopContext.value.setValue({
+                type: "listOf",
+                configFieldId: "list1", 
+                index: 1, 
+                locator: {
+                    type: "single",
+                    configFieldId: "field1InList1"
+                }
+            }, "test"); 
```
oh I imagined we could do something like:
```
workshopContext.value.list1[1].field1InList1.set("test")
```
using mapped types in typescript to map the config into value

faye-tan replied:
Took a while but got this working, for all variable types even structs. Please see updated `Example.tsx`

mduczek reviewed (commented):
nice! this looks promising!

faye-tan commented on src/useWorkshopContext.ts:
```diff
@@ -0,0 +1,107 @@
+
+import React from "react";
+import { IConfigDefinitionFields } from "./types/configDefinition";
+import { IAsyncStatus, asyncStatusLoading, asyncStatusLoaded } from "./types/loadingState";
```
will change

mduczek commented on src/types/variableTypes.ts:
```diff
@@ -0,0 +1,81 @@
+    fieldType: IStructVariableFieldType;
+}
+
+export interface IVariableType_ObjectSet {
+    type: "objectSet";
+}
```
given that we won't support objectSets initially, let's comment it out for now ?
or do we support it somehow?

faye-tan replied:
Discussed previously, but we will support objectSets defined using an objectTypeId and up to 10,000 primaryKeys.

mduczek commented on src/types/loadingState.ts:
```diff
@@ -0,0 +1,104 @@
+/**
+ * Final composite 'async' loading types. 
+ */
+export type IAsyncStatus<V, E = unknown> = 
```
```suggestion
export type IAsyncValue<V, E = unknown> = 
```

mduczek commented on src/example/Example.tsx:
```diff
@@ -0,0 +1,52 @@
+        // set values 
+        loadedContext.output_boolean_field.setValue(asyncStatusLoaded(true)); // input param must be bool or compiler will err
+        loadedContext.listOfConfigFields[0]["listOf-in-listOf"][0]["output-boolean-list-in-nested-list"].setValue(asyncStatusLoaded([true, false])); // input param must be bool[] or compiler will err
+        loadedContext["output-struct-field"].setValue( // TODO this is not working 100% yet 
```
if it's not working could we comment it out so we can release the first version without it?

mduczek commented on src/types/workshopContext.ts:
```diff
@@ -0,0 +1,56 @@
+import { IVariableType, IStructVariableFieldTypes } from "./variableTypes";
+
+export interface ExecutableEvent {
+    executeEvent: () => void;
```
even if we don't immediately implement different states let's use the target API
```suggestion
    executeEvent: IAsyncLoaded<() => void>;
```

faye-tan replied:
The API for executing an event [here](https://github.palantir.build/foundry/forge/blob/develop/packages/workshop/object-analytics-config-spec/src/api/types.ts#L100) is 
```
execute: (mouseEvent: IMouseEvent | React.KeyboardEvent | undefined) => boolean;
```

So in the plugin should it be this instead? 
```
executeEvent: (mouseEvent: IMouseEvent | React.KeyboardEvent | undefined) => boolean;
```

mduczek commented on src/example/Example.tsx:
```diff
@@ -0,0 +1,52 @@
+        // get values 
+        const inputArrayNumberVariable = loadedContext.inputArrayNumberField; // type is number[] | undefined
+        const v2 = loadedContext.listOfConfigFields[0].inputStringFieldInsideListOf; // type is string | undefined 
+        const v3 = loadedContext["input-struct-field"]; 
```
let's use fieldNames that can be accessed with a "." operator so let's use camelCase for fieldNames think it will make it look way better

faye-tan commented on src/useWorkshopContext.ts:
```diff
@@ -0,0 +1,126 @@
+                    }
+                }
+            }
+            return prevConfigDefinitionField;
```
reworked - workshop passes a tree of values and the plugin will simply accept it 
```
/**
   * Entire tree of values is passed from Workshop.
   */
  const handleValueChangeFromWorkshop = (
    newConfigValues: IConfigValueMap,
    iframeWidgetIdFromWorkshop: string
  ) => {
    // Only set map of config values if iframeWidgetId received from Workshop matches saved iframeWidgetId
    if (iframeWidgetIdFromWorkshop === iframeWidgetId) {
      setConfigValues(newConfigValues);
    }
  };
```

faye-tan commented on README.md:
```diff
@@ -1,2 +1,148 @@
-# workhop-iframe-react-app-plugin
-Workshop plugin to enable bidirectional communication Workshop and any custom React app
+# Workshop Iframe React app plugin 
```
Can polish this up in a follow up PR, this is first draft

mduczek commented on README.md:
```diff
@@ -1,2 +1,158 @@
+   During development when your React app is not iframed in Workshop, its not receiving any values and as such it would make development difficult if all you had to work with were a forever loading context object or a loaded context object with null values. Allowing you to provide default values when defining the config that gets translated to the context object returned by `useWorkshopContext` h ...
+
+3. Why is each value inside the context object returned by `useWorkshopContext` wrapped in with an async loading state?
+   Workshop's variables each have an async loading state, as variables could come from asynchronous origins, such as a function that takes 10 seconds to return a value. Having a 1:1 match between the types in the context object and Workshop means that the two can be in sync concerning the state of the shared variable values. If a variable in Workshop goes into a loading state or fails to load, re ...
```
```suggestion
   Workshop's variables each have an async loading state, as variables could come from asynchronous execution, such as a function that takes 10 seconds to return a value. Having a 1:1 match between the types in the context object and Workshop means that the two have a consistent view of variable values. If a variable in Workshop goes into a loading state or fails to load, this async state is passed to the iframed React app allowing you to decide how to handle cases where one of the parameters is not available or currently might be re-loading. For example when implementing a custom submission form, you might want to disable the submission button if some of the inputs to your form hasn't loaded yet or are currently re-loading.
```

mduczek commented on README.md:
```diff
@@ -1,2 +1,158 @@
+3. Why is each value inside the context object returned by `useWorkshopContext` wrapped in with an async loading state?
+   Workshop's variables each have an async loading state, as variables could come from asynchronous origins, such as a function that takes 10 seconds to return a value. Having a 1:1 match between the types in the context object and Workshop means that the two can be in sync concerning the state of the shared variable values. If a variable in Workshop goes into a loading state or fails to load, re ...
+
+## License
+
+TODO
```
raising this :)

faye-tan replied:
I wasn't sure what to put here between MIT and Apache license. Going to ask about this. The README for https://github.palantir.build/devtools/open-source also says: 

If your Java project makes use of patents we hold and/or patents we intend to file (or are filed), please use [BSD-3](https://opensource.org/licenses/BSD-3-Clause) with the legal-provided PATENTS file.

mduczek commented on src/types/messages.ts:
```diff
@@ -0,0 +1,105 @@
+import { ILocator } from "./locator";
+import { IVariableToSet } from "./variableValues";
+
+export enum MESSAGE_TYPES_TO_WORKSHOP {
```
I think it would be good to create an `internal` folder for things that are internal but we still want to export them to use in Workshop code

mduczek commented on src/transformConfigToWorkshopContext.ts:
```diff
@@ -0,0 +1,366 @@
+} from "./types/workshopContext";
+import { assertNever, sendMessageToWorkshop } from "./utils";
+
+// Currently, object sets are limited to 10000 objects. This limit will be removed when osdk supports materializing objects from temporary ObjectRids.
```
```suggestion
// Currently, object sets are limited to 10000 objects. This limit will be removed when osdk supports loading objectSets from temporary ObjectSetRids.
```

mduczek commented on src/transformConfigToWorkshopContext.ts:
```diff
@@ -0,0 +1,366 @@
+): IWorkshopContext<T> {
+  // Recursively traverses the value map and updates it when given
+  // a value and valueLocator path through the value map tree.
+  const createNewConfigValueMapWithValueChange = (
```
nit extract to a function

mduczek commented on src/transformConfigToWorkshopContext.ts:
```diff
@@ -0,0 +1,366 @@
+
+  // Before setting objectSet information in the context's value map, extract the primaryKeys,
+  // which is OSDK's preffered format to load objects with
+  const maybeExtractPrimaryKeysFromVariableTypeToValueValue = <
```
nit extract to a function

mduczek commented on src/transformConfigToWorkshopContext.ts:
```diff
@@ -0,0 +1,366 @@
+
+  // Before sending objectSet information to Workshop, extract the objectRids,
+  // which is Workshop's locator format to load objects with
+  const maybeExtractObjectRidsFromVariableTypeToValueTypeToSet = <
```
nit extract to a function

mduczek commented on src/transformConfigToWorkshopContext.ts:
```diff
@@ -0,0 +1,366 @@
+  };
+
+  // Returns a callback to set a context field as "loaded" with a value
+  const createSetLoadedValueCallback =
```
nit extract to a function

mduczek commented on src/transformConfigToWorkshopContext.ts:
```diff
@@ -0,0 +1,366 @@
+    };
+
+  // Returns a callback to set a context field as "reloading" with a value
+  const createSetReloadingValueCallback =
```
nit extract to a function

mduczek commented on src/transformConfigToWorkshopContext.ts:
```diff
@@ -0,0 +1,366 @@
+    };
+
+  // Returns a callback to set a context field as "loading"
+  const createSetLoadingCallback = (valueLocator: ILocator) => () => {
```
nit extract to a function

mduczek commented on src/transformConfigToWorkshopContext.ts:
```diff
@@ -0,0 +1,366 @@
+  };
+
+  // Returns a callback to set a context field as "failed to load"
+  const createSetFailedWithErrorCallback =
```
nit extract to a function

mduczek commented on src/transformConfigToWorkshopContext.ts:
```diff
@@ -0,0 +1,366 @@
+    };
+
+  // Returns a callback to execute an event in Workshop
+  const createExecuteEventCallback =
```
nit extract to a function

mduczek commented on src/transformConfigToWorkshopContext.ts:
```diff
@@ -0,0 +1,366 @@
+      case "single":
+        switch (field.fieldValue.type) {
+          case "event": {
+            const locator: ILocator =
+              opts == null
+                ? { type: "single", configFieldId: fieldId }
+                : opts.createLocatorInListCallback({
+                    type: "single",
+                    configFieldId: fieldId,
+                  });
```
this is the same across the two branches, let's extract this above the inner switch

mduczek commented on src/transformConfigToWorkshopContext.ts:
```diff
@@ -0,0 +1,366 @@
+        }
+      case "listOf":
+        if (configValues[fieldId] != null && configValues[fieldId].type === "listOf") {
+          const createLocator =
```
extract as function `getCreateLocator(configFieldId: string, index: number)`

mduczek commented on src/example/Example.tsx:
```diff
@@ -0,0 +1,159 @@
+    booleanField.fieldValue;
+  const numberFieldValue: IAsyncLoaded<number> | undefined =
+    numberField.fieldValue;
+  const dateFieldValue: IAsyncLoaded<Date> | undefined = dateField.fieldValue;
```
```suggestion
  const dateFieldValue: IAsyncLoaded<string> | undefined = dateField.fieldValue;
```

mduczek commented on src/example/Example.tsx:
```diff
@@ -0,0 +1,159 @@
+  /**
+   * Examples of ListOf config. These are accessed by index.
+   */
+  listOfFields.forEach((listOfLayer, index) => {
```
```suggestion
  listOfFields.forEach((listItem, index) => {
```

mduczek commented on src/example/Example.tsx:
```diff
@@ -0,0 +1,159 @@
+    );
+
+    // Can have nested listOf configs
+    listOfLayer.nestedListOfField.forEach((nestedListOfLayer, nestedIndex) => {
```
```suggestion
    listItem.nestedListOfField.forEach((nestedListItem, nestedIndex) => {
```

faye-tan commented on src/useWorkshopContext.ts:
```diff
@@ -0,0 +1,143 @@
+  // Config was not accepted by workshop, return failed
+  if (isConfigRejectedByWorkshop && insideIframe) {
+    return asyncStatusFailed(
+      "Workshop rejected the config definition. This is likely due to an API break, as the Workshop iframe's existing saved config is not a subset of the provided config."
```
Ah yes, will update message.

faye-tan commented on src/types/workshopContext.ts:
```diff
@@ -0,0 +1,203 @@
+   * then please pass the mouse event too. This allows the framework to
+   * choose to open things in new tabs if e.g. the CTRL button is pressed.
+   *
+   * Returns false if the event was not able to be executed (e.g. the CTRL button
+   * was held down, but there was no way to execute the transition in a new tab,
+   * and so it was deemed safer to not execute it at all, instead of risk overwriting
+   * the users state in the existing tab).
```
removed 😅

faye-tan commented on src/types/workshopContext.ts:
```diff
@@ -0,0 +1,203 @@
+    : T extends { type: "number" }
+    ? number
+    : T extends { type: "date" } | { type: "timestamp" }
+    ? Date
```
Yes! I had not updated this Friday in this repo but had changed it in my OSDK app so that the PR preview would not be broken. Moving my changes over to this repo.

faye-tan commented on src/index.ts:
```diff
@@ -0,0 +1,19 @@
+ */
+
+export { useWorkshopContext } from "./useWorkshopContext";
+export * from "./internal";
```
exports types Workshop needs

mduczek commented on src/example/Example.tsx:
```diff
@@ -0,0 +1,163 @@
+} from "../types/loadingState";
+import { IWorkshopContext } from "../types/workshopContext";
+import { useWorkshopContext } from "../";
+import { ObjectSetLocators } from "../internal";
```
since we expect builders to use it, it shouldn't be in internal types

mduczek commented on src/internal/configDefinition.ts:
```diff
@@ -0,0 +1,60 @@
+ * The config definition consists of a readonly list of config fields.
+ * Ensure that there are no duplicate fieldIds in each of the config fields.
+ */
+export type IConfigDefinition = readonly IConfigDefinitionField[];
```
I think those aren't internal types, builders will use them to define config

mduczek commented on src/internal/index.ts:
```diff
@@ -0,0 +1,23 @@
+ */
+
+// Exported for Workshop usage
+export * from "./configDefinition";
```
this will mostly likely be used by builders?

mduczek commented on src/internal/index.ts:
```diff
@@ -0,0 +1,23 @@
+// Exported for Workshop usage
+export * from "./configDefinition";
+export * from "./configValues";
+export * from "./locator";
+export * from "./messages";
```
those two are definitely internal

mduczek commented on src/types/loadingState.ts:
```diff
@@ -0,0 +1,193 @@
+/**
+ * Helper function for creating "not started" async loading state.
+ */
+export function asyncStatusNotStarted(): IAsyncValue_NotStarted {
```
```suggestion
export function asyncValueNotStarted(): IAsyncValue_NotStarted {
```

mduczek commented on src/types/loadingState.ts:
```diff
@@ -0,0 +1,193 @@
+ * @type E: The error type
+ * @returns true only if the asyncloaded state is "not started".
+ */
+export function isAsyncStatus_NotStarted<V, E>(
```
```suggestion
export function isAsyncValue_NotStarted<V, E>(
```

mduczek commented:
should this still be a date?
https://github.com/faye-tan/workshop-iframe-react-app-plugin/pull/1#discussion_r1817900851

faye-tan commented:
> should this still be a date? [#1 (comment)](https://github.com/faye-tan/workshop-iframe-react-app-plugin/pull/1#discussion_r1817900851)

For `VariableTypeToValueType` I have changed the type to `string` for variable type date and `string[]` for variable type date-list. This is the mapped type that maps from a variable type to the value type that actually holds the date. For `VariableTypeToValueTypeToSet` on the other hand, it should still be type Date and Date[] for variable types date and date-list, as this is the type for values in setter methods and I think it makes more sense for a consumer to call 

```
dateField.setLoadedValue(value: Date)
```
over 
```
dateField.setLoadedValue(value: string)
```


Added comments to clarify.

mduczek reviewed (approved):
let's release the first version!
we can fix issues as we go before major version :)
